### PR TITLE
feat(datasets): added ls-files command

### LINKED ISSUE #446: cli: add ls-files command
Add ls-files command to list all files within a dataset.

Example: 
```console
$ renku dataset ls-files my-dataset
ADDED                AUTHORS    FILENAME
-------------------  ---------  ----------
2019-02-28 16:48:09             addme
2019-02-28 16:49:02             file1
2019-02-28 16:49:02             file2
2019-02-28 16:49:02             file3
```

Sorting: by added date

related #418

### PR DESCRIPTION
**Pull request by jsam**
This PR adds ls-files command which will list all files with a given set of filters.

Fixes #446

### CODE REVIEW

File: renku/cli/dataset.py [hunk 1]
```diff
@@ -84,6 +84,7 @@
 from ._client import pass_local_client
 from ._echo import progressbar
 from ._format.datasets import FORMATS as DATASETS_FORMATS
+from ._format.ls_files import FORMATS as LS_FILES_FORMATS
```
Comment (jirikuncar) [inline renku/cli/dataset.py]:
```suggestion
from ._format.dataset_files import FORMATS as DATASET_FILES_FORMATS
```

File: renku/cli/dataset.py [hunk 2]
```diff
@@ -157,6 +158,39 @@ def add(client, name, urls, nocopy, relative_to, target, force):
...
+    help='Choose an output format.'
+)
+@pass_local_client(clean=False, commit=False)
+def ls_files(client, format, exclude, include, authors, dataset):
+    """List files in dataset."""
+    LS_FILES_FORMATS[format](
+        client,
+        dataset_name=dataset,
+        authors=authors,
+        include=include,
```
Comment (jirikuncar) [inline renku/cli/dataset.py]: please move the `include` and `exclude` logic out from the formatter.

Comment (jsam) [reply renku/cli/dataset.py]: I agree. Was thinking into extracting that part so I can reuse it on #447. Can you point me where should I put that part of code?

Comment (jirikuncar) [reply renku/cli/dataset.py]: I was thinking about something simple likeL
```python
records = _filter(datasets, dataset_name, authors, include, exclude)
LS_FILES_FORMATS[format](client, records)
```

**Comment by jsam** [comment]
1. Reply ls_files.py to dataset_files.py

File: renku/cli/dataset.py [hunk 3]
```diff
@@ -157,6 +161,98 @@ def add(client, name, urls, nocopy, relative_to, target, force):
...
+@click.option(
+    '--dataset', multiple=True, help='Filter files in specific dataset.'
+)
+@click.option(
+    '--authors',
+    help='Filter files which where authored by specific authors. '
+    'Multiple authors are specified by comma.'
+)
+@click.option(
+    '--include',
```
Comment (jirikuncar) [inline renku/cli/dataset.py]: can you add short option `-I`

File: renku/cli/dataset.py [hunk 4]
```diff
@@ -157,6 +161,98 @@ def add(client, name, urls, nocopy, relative_to, target, force):
...
+    default='*',
```
Comment (jirikuncar) [inline renku/cli/dataset.py]:
```suggestion
    default=None,
```

File: renku/cli/dataset.py [hunk 5]
```diff
@@ -157,6 +161,98 @@ def add(client, name, urls, nocopy, relative_to, target, force):
...
+    multiple=True,
+    help='Include files matching given pattern.'
+)
+@click.option(
+    '--exclude', multiple=True, help='Exclude files matching given pattern.'
```
Comment (jirikuncar) [inline renku/cli/dataset.py]: can you add short option `-X`?

File: renku/cli/dataset.py [hunk 6]
```diff
@@ -157,6 +161,98 @@ def add(client, name, urls, nocopy, relative_to, target, force):
...
+        dataset_names=dataset,
+        authors=authors,
+        include=include,
+        exclude=exclude
+    )
+
+    DATASET_FILES_FORMATS[format](client, records)
+
+
+def _include_exclude(file_path, include, exclude=None):
```
Comment (jirikuncar) [inline renku/cli/dataset.py]:
```suggestion
def _include_exclude(file_path, include=None, exclude=None):
```

File: renku/cli/dataset.py [hunk 7]
```diff
@@ -157,6 +161,98 @@ def add(client, name, urls, nocopy, relative_to, target, force):
...
+    :param exclude: Tuple containing patterns to which exclude from result.
+    """
+    filename = Path(file_path).name
+
+    for pattern in exclude:
+        if fnmatch(filename, pattern):
+            return False
+
+    found = False
+    for pattern in include:
```
Comment (jirikuncar) [inline renku/cli/dataset.py]: it should be enough if it's only in one include pattern:

```python
if include is not None:
    for pattern in include:
       if fnmatch(filename, pattern):
            return True
    return False
return True

File: renku/cli/dataset.py [hunk 8]
```diff
@@ -157,6 +161,106 @@ def add(client, name, urls, link, relative_to, target, force):
...
+
+
+def _include_exclude(file_path, include=None, exclude=None):
+    """Check if file matches one of include filters and not in exclude filter.
+
+    :param file_path: Path to the file.
+    :param include: Tuple containing patterns to which include from result.
+    :param exclude: Tuple containing patterns to which exclude from result.
+    """
+    filename = Path(file_path).name
```
Comment (jirikuncar) [inline renku/cli/dataset.py]: we should match the full path

File: tests/cli/test_datasets.py [hunk 9]
```diff
@@ -309,3 +310,146 @@ def test_dataset_add_with_copy(tmpdir, runner, project, client):
...
+def test_datasets_ls_files_tabular_patterns(tmpdir, runner, project):
+    """Test listing of data within dataset with include/exclude filters."""
+    # create a dataset
+    result = runner.invoke(cli.cli, ['dataset', 'create', 'my-dataset'])
+    assert result.exit_code == 0
+
+    # create some data
+    paths = []
+    for i in range(3):
+        new_file = tmpdir.join('file_{0}'.format(i))
```
Comment (jirikuncar) [inline tests/cli/test_datasets.py]: we should test subdirectories too

File: renku/cli/dataset.py (hunk 8 above)
Comment (jsam) [reply renku/cli/dataset.py]: We discussed this and you said that we want just files without a path. :confused: So which one in the end?

File: renku/cli/dataset.py [hunk 10]
```diff
@@ -157,6 +161,106 @@ def add(client, name, urls, link, relative_to, target, force):
...
+
+    if isinstance(authors, list) or isinstance(authors, tuple):
+        authors = set(authors)
+
+    records = []
+    for path_, dataset in datasets.items():
+        if dataset.name in dataset_names or not dataset_names:
+            for file_ in dataset.files.values():
+                file_.dataset = dataset.name
+                match = _include_exclude(file_.filename, include, exclude)
```
Comment (jirikuncar) [inline renku/cli/dataset.py]: more or less it should be `(file.__reference__.parent / file.path).relative_to(client.path)`

File: renku/cli/dataset.py [hunk 11]
```diff
@@ -157,6 +160,106 @@ def add(client, name, urls, link, relative_to, target, force):
...
+def _include_exclude(file_path, include=None, exclude=None):
+    """Check if file matches one of include filters and not in exclude filter.
+
+    :param file_path: Path to the file.
+    :param include: Tuple containing patterns to which include from result.
+    :param exclude: Tuple containing patterns to which exclude from result.
+    """
+    if exclude is not None and exclude:
+        for pattern in exclude:
+            if fnmatch(str(file_path), pattern):
```
Comment (jirikuncar) [inline renku/cli/dataset.py]:
```suggestion
            if file_path.match(pattern):
```

https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.match

File: renku/cli/dataset.py [hunk 12]
```diff
@@ -157,6 +160,106 @@ def add(client, name, urls, link, relative_to, target, force):
...
+                return False
+
+    if include is not None and include:
+        for pattern in include:
+            if fnmatch(str(file_path), pattern):
```
Comment (jirikuncar) [inline renku/cli/dataset.py]:
```suggestion
            if file_path.match(pattern):
```

File: renku/models/datasets.py [hunk 13]
```diff
@@ -16,9 +16,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Model objects representing datasets."""
-
```
Comment (jirikuncar) [inline renku/models/datasets.py]:
```suggestion

```

File: tests/cli/test_datasets.py [hunk 14]
```diff
@@ -237,7 +239,7 @@ def test_relative_git_import_to_dataset(tmpdir, runner, project, client):
 
 
 def test_dataset_add_with_link(tmpdir, runner, project, client):
-    """Test adding data to dataset with --no-copy flag."""
+    """Test adding data to dataset with --link flag."""
     import os
```
Comment (jirikuncar) [inline tests/cli/test_datasets.py]:
```suggestion
```

File: renku/cli/dataset.py [hunk 15]
```diff
@@ -157,6 +158,106 @@ def add(client, name, urls, link, relative_to, target, force):
...
+        if dataset.name in dataset_names or not dataset_names:
+            for file_ in dataset.files.values():
+                file_.dataset = dataset.name
+
+                path_ = file_.full_path.relative_to(client.path)
+                match = _include_exclude(path_, include, exclude)
+
+                if authors:
+                    match = match and authors.issubset(
+                        set([author.name for author in file_.authors])
```
Comment (jirikuncar) [inline renku/cli/dataset.py]:
```suggestion
                        {author.name for author in file_.authors}
```

**Pull request merged**